package/src/web_voice_processor.ts

```diff
@@ -168,88 +71,215 @@ export class WebVoiceProcessor {
    * @param durationMs the duration of the recording, in milliseconds
    * @return the data in Blob format, wrapped in a promise
    */
-
   public async audioDump(durationMs: number = 3000): Promise<Blob> {
-    if (this._audioDumpPromise !== null) {
-      return Promise.reject('Audio dump already in progress');
-    }
-
-    this._downsamplingWorker.postMessage({
-      command: 'start_audio_dump',
-      durationMs: durationMs,
+    const audioDumpEngine = new AudioDumpEngine();
+    this.subscribe(audioDumpEngine);
+    return new Promise<Blob>(resolve => {
+      // @ts-ignore
+      this.audioDumpEngine = audioDumpEngine;
+      setTimeout(() => {
+        this.unsubscribe(audioDumpEngine);
+        resolve(audioDumpEngine.onend());
+      }, durationMs);
     });
+  }
 
-    this._audioDumpPromise = new Promise<Blob>((resolve, reject) => {
-      this._audioDumpResolve = resolve;
-      this._audioDumpReject = reject;
-    });
+  /**
+   * Subscribe an engine. A subscribed engine will receive audio frames via
+   * `.postMessage({command: 'process', inputFrame: inputFrame})`.
+   * @param engine The engine to unsubscribe.
+   */
+  public subscribe(engine: PvEngine): void {
+    if (engine.worker) {
+      if (engine.worker.postMessage && typeof engine.worker.postMessage === 'function') {
+        this._engines.add(engine);
+      } else {
+        throw new Error("Engine must have a 'onmessage' handler.");
+      }
+    } else {
+      if (engine.postMessage && typeof engine.postMessage === 'function') {
+        this._engines.add(engine);
+      } else if (engine.onmessage && typeof engine.onmessage === 'function') {
+        this._engines.add(engine);
+      } else {
+        throw new Error("Engine must have a 'onmessage' handler.");
```

Comment:
I'm confused as to whether an Engine needs `onmessage` or `postMessage` or both to go be accepted.

Reply:
this the user should only think that a onmessage is required. Internally, an object just has an onmessage so that is required. postMessage exists when there is a worker. A worker instance may not have a onmessage yet but postmessage is always there. so in general:
- main: `engine` -> `onmessage`
- worker: `engine` -> `worker.postmessage` -> (inside worker)`onmessage` -> (inside worker) `postmessage` -> (back to engine) `onmessage`

Reply:
cool, got it